modules/core/src/main/java/org/apache/ignite/IgniteSemaphore.java

```diff
@@ -244,6 +247,16 @@ public boolean tryAcquire(int permits, long timeout, TimeUnit unit)
      */
     public void acquire(int permits) throws IgniteInterruptedException;
 
+    /**
+     * Acquires the given semaphore, executes the given callable and schedules the release of permits asynchronously
+     *
+     * @param callable the callable to execute
+     * @param numPermits the number of permits to acquire
+     * @throws Exception if the callable throws an exception
+     */
+    public <T> IgniteFuture<T> acquireAndExecute(IgniteCallable<IgniteFuture<T>> callable,
+                                                 int numPermits) throws Exception;
+
```

Comment:
I'm not sure I understand why callable returns `IgniteFuture<T>` rather than just `T`. I saw a comment in the ticket, but it doesn't really provide justification. It probably makes sense to have a separate thread on the dev list regarding this - we should have a discussion if there are different opinions.

Reply:
Agreed, fixed the same